pkg/utils/etcdutil/etcdutil.go

```diff
@@ -350,3 +353,227 @@ func InitOrGetClusterID(c *clientv3.Client, key string) (uint64, error) {
+		log.Warn("watched key doesn't exist in watch loop when loading", zap.String("name", lw.name), zap.String("key", lw.key), zap.Error(err))
+	}
+	lw.isLoadedCh <- err
+	return watchStartRevision
+}
+
+func (lw *LoopWatcher) watch(ctx context.Context, revision int64) (nextRevision int64, err error) {
+	watcher := clientv3.NewWatcher(lw.client)
+	defer watcher.Close()
+
+	for {
+	WatchChan:
+		opts := append(lw.opts, clientv3.WithRev(revision))
+		watchChan := watcher.Watch(ctx, lw.key, opts...)
+		select {
+		case <-ctx.Done():
+			return revision, nil
+		case <-lw.forceLoadCh:
+			revision, err = lw.load()
+			if err != nil {
+				log.Warn("force load key failed in watch loop", zap.String("name", lw.name),
+					zap.String("key", lw.key), zap.Error(err))
+			}
+			goto WatchChan
+		case wresp := <-watchChan:
+			if wresp.CompactRevision != 0 {
+				log.Warn("required revision has been compacted, use the compact revision in watch loop",
+					zap.Int64("required-revision", revision),
+					zap.Int64("compact-revision", wresp.CompactRevision))
+				revision = wresp.CompactRevision
+				goto WatchChan
+			} else if wresp.Err() != nil { // wresp.Err() contains CompactRevision not equal to 0
+				log.Error("watcher is canceled in watch loop",
+					zap.Int64("revision", revision),
+					errs.ZapError(errs.ErrEtcdWatcherCancel, wresp.Err()))
+				return revision, wresp.Err()
+			}
+			for _, event := range wresp.Events {
+				switch event.Type {
+				case clientv3.EventTypePut:
+					if err := lw.putFn(event.Kv); err != nil {
+						log.Error("put failed in watch loop", zap.String("name", lw.name),
+							zap.String("key", lw.key), zap.Error(err))
+					}
+				case clientv3.EventTypeDelete:
+					if err := lw.deleteFn(event.Kv); err != nil {
+						log.Error("delete failed in watch loop", zap.String("name", lw.name),
+							zap.String("key", lw.key), zap.Error(err))
+					}
+				}
+			}
+			revision = wresp.Header.Revision + 1
+		}
+	}
+}
+
+func (lw *LoopWatcher) load() (nextRevision int64, err error) {
+	startKey := lw.key
+	limit := lw.loadBatchSize
+	if limit <= 0 {
+		limit = defaultLoadBatchSize
+	}
+	for {
+		opts := append(lw.opts, clientv3.WithSort(clientv3.SortByKey, clientv3.SortAscend), clientv3.WithLimit(limit))
+		resp, err := EtcdKVGet(lw.client, startKey, opts...)
+		failpoint.Inject("delayLoad", func(val failpoint.Value) {
+			if sleepIntervalSeconds, ok := val.(int); ok && sleepIntervalSeconds > 0 {
+				time.Sleep(time.Duration(sleepIntervalSeconds) * time.Second)
+			}
+		})
+
+		if err != nil {
+			log.Error("load failed in watch loop", zap.String("name", lw.name),
+				zap.String("key", lw.key), zap.Error(err))
+			return 0, err
+		}
+		for _, item := range resp.Kvs {
+			err = lw.putFn(item)
+			if err != nil {
+				log.Error("put failed in watch loop when loading", zap.String("name", lw.name), zap.String("key", lw.key), zap.Error(err))
+				continue
+			}
+		}
+		count := int64(len(resp.Kvs))
+		if count < limit {
```

Comment:
Can we use clientv3.GetResponse.More to judge if we have retrieved all the entries?

Reply:
The etcd comments only says "When passed WithLimit(limit), the number of returned keys is bounded by limit" but it doesn't mean it's done the job if count < limit here. According to another comment "// more indicates if there are more keys to return in the requested range." it seems that more is safer to use.